File: gke_customization/gke_catalog/report/sketch_order_summary/sketch_order_summary.js

// Copyright (c) 2025, Gurukrupa Export and contributors
// For license information, please see license.txt

frappe.query_reports["Sketch Order Summary"] = {
       filters: [
       {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.add_days(frappe.datetime.get_today(), -30),
            reqd: 0
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.get_today(),
            reqd: 0
        },
         {
            fieldname: "branch",
            label: __("Branch"),
            fieldtype: "Link",
            options: "Branch",
           // default: frappe.defaults.get_user_default("branch")
        },
        {
            fieldname: "customer_group",
            label: __("Customer Group"),
            fieldtype: "Link",
            options: "Customer Group"
        },
        {
            fieldname: "customer",
            label: __("Customer"),
            fieldtype: "Link",
            options: "Customer"
        }
    ],

 formatter: function (value, row, column, data, default_formatter) {
    value = default_formatter(value, row, column, data);

    if (column.fieldname === "customer_codes") {
        return "";
    }

    const drill_fields = [
        "sof_draft", "sof_submitted", "sof_cancelled",
        "so_draft", "so_submitted", "so_cancelled"
    ];

    if (!drill_fields.includes(column.fieldname)) {
        return value;
    }

    if (!value || parseInt(value) === 0) {
        return value;
    }

    const params = new URLSearchParams();
    const filter_keys = ["from_date", "to_date", "branch"];
    filter_keys.forEach(key => {
        const filter_value = frappe.query_report.get_filter_value(key);
        if (filter_value) {
            params.append(key, filter_value);
        }
    });

if (column.fieldname === "sof_draft" || column.fieldname === "sof_cancelled") {
    params.append("docstatus", column.fieldname === "sof_draft" ? 0 : 2);

    const from = frappe.query_report.get_filter_value("from_date");
    const to = frappe.query_report.get_filter_value("to_date");
    if (from && to) {
        params.append("order_date", JSON.stringify(["Between", [from, to]]));
    }

    if (data.customer_codes) {
        const customers = data.customer_codes
            .split(',')
            .map(c => c.trim())
            .filter(Boolean);

        if (customers.length > 0) {
            params.append("customer_code", JSON.stringify(["in", customers]));
        }
    }

    const url = `/app/sketch-order-form?${params.toString()}`;
    return `<a href="${url}" target="_blank" style="text-decoration: inherit; color: inherit;">${value}</a>`;
}


    const report_name = "Sketch Order Detail";

    if (data.customer_codes) {
        params.append("customer", data.customer_codes);
    }

 //   if (column.fieldname === "sof_submitted") params.append("docstatus", 1);
    if (column.fieldname === "so_draft") params.append("docstatus", 0);
    if (column.fieldname === "so_submitted") params.append("docstatus", 1);
    if (column.fieldname === "so_cancelled") params.append("workflow_state", "Cancelled");

    const url = `/app/query-report/${encodeURIComponent(report_name)}?${params.toString()}`;
    return `<a href="${url}" target="_blank" style="text-decoration: inherit; color: inherit;">${value}</a>`;
},

onload: function(report) {
     report.page.add_inner_button(__("Clear Filter"), function () {
            report.filters.forEach(function (filter) {
                let field = report.get_filter(filter.fieldname);
                if (field.df.fieldtype === "MultiSelectList") {
                    field.set_value([]);
                } else if (field.df.default) {
                    field.set_value(field.df.default);
                } else {
                    field.set_value("");
                }
            });
        });
},
// formatter: function (value, row, column, data, default_formatter) {
//     // Only make clickable links for non-zero values
//     if (value > 0) {
//         return `<a href="/app/query-report/Sketch Order Detail">${value}</a>`;
//     }

//     return default_formatter(value, row, column, data);
// }};
}

File: gke_customization/gke_catalog/report/sketch_order_summary/sketch_order_summary.py
# Copyright (c) 2025, Gurukrupa Export and contributors
# For license information, please see license.txt

import frappe

def execute(filters=None):
    filters = filters or {}
    data = get_data(filters)
    columns = get_columns()

    if data:
        total_row = {
            "customer_category": "Total",
            "sof_draft": sum(row["sof_draft"] for row in data),
            "sof_submitted": sum(row["sof_submitted"] for row in data),
            "sof_cancelled": sum(row["sof_cancelled"] for row in data),
            "so_draft": sum(row["so_draft"] for row in data),
            "so_submitted": sum(row["so_submitted"] for row in data),
            "so_cancelled": sum(row["so_cancelled"] for row in data),
            "customers": "",  
        }
        data.append(total_row)

    return columns, data
    # return get_columns(), get_data(filters)

def get_columns():
    return [
        {"label": "Customer Group", "fieldname": "customer_category", "fieldtype": "Data", "width": 150},
        # {"label": "Customer Code", "fieldname": "customer_codes", "fieldtype": "Data", "width": 150},
        # {"label": "Customers", "fieldname": "customers", "fieldtype": "Data", "width": 300},
        {"label": "Sketch OF Draft", "fieldname": "sof_draft", "fieldtype": "Int", "width": 170},
        {"label": "Sketch OF Submitted", "fieldname": "sof_submitted", "fieldtype": "Int", "width": 170},
        {"label": "Sketch OF Cancelled", "fieldname": "sof_cancelled", "fieldtype": "Int", "width": 170},
        {"label": "Sketch Order Draft", "fieldname": "so_draft", "fieldtype": "Int", "width": 170},
        {"label": "Sketch Order Submitted", "fieldname": "so_submitted", "fieldtype": "Int", "width": 190},
        {"label": "Sketch Order Cancelled", "fieldname": "so_cancelled", "fieldtype": "Int", "width": 190},
    ]

def get_data(filters):
    conditions = []

    if filters.get("branch"):
        conditions.append("sof.branch = %(branch)s")

    if filters.get("from_date"):
        conditions.append("sof.order_date >= %(from_date)s")

    if filters.get("to_date"):
        conditions.append("sof.order_date <= %(to_date)s")

    if filters.get("customer_group"):
        conditions.append("cg.customer_group_name = %(customer_group)s")
        
    if filters.get("customer"):
        conditions.append("c.name = %(customer)s")

    condition_sql = " AND ".join(conditions)
    if condition_sql:
        condition_sql = "WHERE " + condition_sql

    query = f"""
        SELECT 
            CASE 
                WHEN cg.customer_group_name IN ('Groups', 'Individual', 'Common Party', 'Wholesale') THEN 'Wholesale'
                WHEN cg.customer_group_name = 'Internal' THEN 'Internal'
                WHEN cg.customer_group_name = 'Retail' THEN 'Retail'
                ELSE 'Other-Groups'
            END AS customer_category,

            COUNT(DISTINCT CASE WHEN sof.docstatus = 0 THEN sof.name END) AS sof_draft,
            COUNT(DISTINCT CASE WHEN sof.docstatus = 1 THEN sof.name END) AS sof_submitted,
            COUNT(DISTINCT CASE WHEN sof.docstatus = 2 THEN sof.name END) AS sof_cancelled,

            COUNT(DISTINCT CASE WHEN so.docstatus = 0 THEN so.name END) AS so_draft,
            COUNT(DISTINCT CASE WHEN so.docstatus = 1 THEN so.name END) AS so_submitted,
            COUNT(DISTINCT CASE WHEN so.workflow_state = 'Cancelled' THEN so.name END) AS so_cancelled,
            
            GROUP_CONCAT(DISTINCT c.customer_name SEPARATOR ', ') AS customers,
            GROUP_CONCAT(DISTINCT c.name SEPARATOR ',') AS customer_codes,
            COUNT(DISTINCT c.customer_name) AS customer_count

        FROM `tabSketch Order Form` sof
        LEFT JOIN `tabSketch Order` so ON so.sketch_order_form = sof.name
        LEFT JOIN `tabCustomer` c ON sof.customer_code = c.name
        LEFT JOIN `tabCustomer Group` cg ON c.customer_group = cg.name
        {condition_sql}
        GROUP BY customer_category
        ORDER BY customer_category
    """

    return frappe.db.sql(query, filters, as_dict=True)




            # SUM(CASE WHEN sof.docstatus = 0 THEN 1 ELSE 0 END) AS sof_draft,
            # COUNT(DISTINCT CASE WHEN sof.docstatus = 1 THEN sof.name END) AS sof_submitted,
            # SUM(CASE WHEN sof.docstatus = 2 THEN 1 ELSE 0 END) AS sof_cancelled,

            # SUM(CASE WHEN so.docstatus = 0 THEN 1 ELSE 0 END) AS so_draft,
            # SUM(CASE WHEN so.docstatus = 1 THEN 1 ELSE 0 END) AS so_submitted,
            # SUM(CASE WHEN so.workflow_state = 'Cancelled' THEN 1 ELSE 0 END) AS so_cancelled,
            
